contracts/Strategy.sol

```diff
@@ -53,80 +143,224 @@ contract Strategy is BaseStrategy {
             uint256 _debtPayment
         )
     {
-        // TODO: Do stuff here to free up any returns back into `want`
-        // NOTE: Return `_profit` which is value generated by all positions, priced in `want`
-        // NOTE: Should try to free up at least `_debtOutstanding` of underlying position
+        (, uint256 lastAccrued, ) = kashiPair.accrueInfo();
+        if (block.timestamp > lastAccrued) {
+            kashiPair.accrue();
+        }
+
+        uint256 assets = estimatedTotalAssets();
+        uint256 wantBal = want.balanceOf(address(this));
+
+        uint256 debt = vault.strategies(address(this)).totalDebt;
+
+        if (assets >= debt) {
+            _debtPayment = _debtOutstanding;
+            _profit = assets.sub(debt);
+
+            uint256 amountToFree = _profit.add(_debtPayment);
+
+            if (amountToFree > 0 && wantBal < amountToFree) {
+                liquidatePosition(amountToFree);
+
+                uint256 newLoose = want.balanceOf(address(this));
+
+                //if we dont have enough money adjust _debtOutstanding and only change profit if needed
+                if (newLoose < amountToFree) {
+                    if (_profit > newLoose) {
+                        _profit = newLoose;
+                        _debtPayment = 0;
+                    } else {
+                        _debtPayment = Math.min(
+                            newLoose.sub(_profit),
+                            _debtPayment
+                        );
+                    }
+                }
+            }
+        } else {
+            //serious loss should never happen but if it does lets record it accurately
+            _loss = debt.sub(assets);
```

Comment:
If we have losses AND debtOutstanding, we do not repay the debt (as _debtPayment will remain to be 0 and profit will be 0 too), vault.report will not try to get want back. 

Is that intended? A lossy scenario might "lock" principal and keep sending losses to the vault.

Reply:
Does this seem more appropriate?

```solidity
    function prepareReturn(uint256 _debtOutstanding)
        internal
        override
        returns (
            uint256 _profit,
            uint256 _loss,
            uint256 _debtPayment
        )
    {
        (, uint256 lastAccrued, ) = kashiPair.accrueInfo();
        if (block.timestamp > lastAccrued) {
            kashiPair.accrue();
        }

        uint256 assets = estimatedTotalAssets();
        uint256 wantBal = balanceOfWant();

        uint256 debt = vault.strategies(address(this)).totalDebt;

        if (assets >= debt) {
            _profit = assets.sub(debt);
        } else {
            //serious loss should never happen but if it does lets record it accurately
            _loss = debt.sub(assets);
        }

        _debtPayment = _debtOutstanding;
        uint256 amountToFree = _debtPayment.add(_profit);

        if (amountToFree > 0 && wantBal < amountToFree) {
            liquidatePosition(amountToFree.sub(wantBal));

            uint256 newLoose = balanceOfWant();

            //if we dont have enough money adjust _debtOutstanding and only change profit if needed
            if (newLoose < amountToFree) {
                if (_profit > newLoose) {
                    _profit = newLoose;
                    _debtPayment = 0;
                } else {
                    _debtPayment = Math.min(
                        newLoose.sub(_profit),
                        _debtPayment
                    );
                }
            }
        }
    }```

Reply:
that would solve what I pointed out, yes. 

is `liquidatePosition` always able to free the requested amountToFree?

If not, you should handle that scenario using returnvalues from liquidatePosition (amountFreed + loss).

Reply:
Also fixed in https://github.com/fp-crypto/kashi-strategy/pull/1/commits/ea17e9d155c17bfcca7f5c5537306c5e52b87cf6